src/base/materialized.ts

```diff
@@ -230,60 +235,143 @@ export class MaterializedChatMessage implements IMaterializedNode {
+			if (text instanceof MaterializedChatMesageImage) {
+				if (text.children.length > 0) {
+					throw new Error('Images cannot have children');
+				}
+				result.push(text);
+				continue;
+			}
 			if (
 				text.lineBreakBefore === LineBreakBefore.Always ||
 				(text.lineBreakBefore === LineBreakBefore.IfNotTextSibling && !isTextSibling)
 			) {
-				if (result.length && !result.endsWith('\n')) {
-					result += '\n';
+				let prev = result[result.length - 1]
+				if (typeof prev === 'string' && result.length && !prev.endsWith('\n')) {
+					result.push('\n');
 				}
 			}
 
-			result += text.text;
+			if (typeof result[result.length - 1] === 'string') {
+				result[result.length - 1] += text.text;
+			} else {
+				result.push(text.text);
+			}
 		}
 
-		return result.trim();
+		return result;
 	});
 
 	public toChatMessage(): ChatMessage {
+		const content = this.text
+			.filter(element => typeof element === 'string')
+			.join('').trim();
+
+		if (this.text.some(element => element instanceof MaterializedChatMesageImage)) {
+			if (this.role !== ChatRole.User) {
+				throw new Error('Only User messages can have images');
+			}
+
+			let prompts: ChatCompletionContentPart[] = this.text.map(element => {
+				if (typeof element === 'string') {
+					return { type: 'text', text: element };
+				} else if (element instanceof MaterializedChatMesageImage) {
+					return {
+						type: 'image_url',
+						image_url: { url: getEncodedBase64(element.imageUrl), detail: element.detail },
+					};
+				} else {
+					throw new Error('Unexpected element type');
+				}
+			});
+
+			return {
+				role: ChatRole.User,
+				content: prompts,
+			};
+		}
+
 		if (this.role === ChatRole.System) {
 			return {
 				role: this.role,
-				content: this.text,
+				content,
 				...(this.name ? { name: this.name } : {}),
 			};
 		} else if (this.role === ChatRole.Assistant) {
 			return {
 				role: this.role,
-				content: this.text,
+				content,
 				...(this.toolCalls ? { tool_calls: this.toolCalls } : {}),
 				...(this.name ? { name: this.name } : {}),
 			};
 		} else if (this.role === ChatRole.User) {
 			return {
 				role: this.role,
-				content: this.text,
+				content,
 				...(this.name ? { name: this.name } : {}),
 			};
 		} else if (this.role === ChatRole.Tool) {
 			return {
 				role: this.role,
-				content: this.text,
+				content,
 				tool_call_id: this.toolCallId,
 			};
 		} else {
 			return {
 				role: this.role,
-				content: this.text,
+				content,
 				name: this.name!,
 			};
 		}
 	}
 }
 
+export class MaterializedChatMesageImage implements IMaterializedNode {
+	constructor(
+		public readonly id: number,
+		public readonly role: ChatRole,
+		public readonly imageUrl: string,
+		public readonly priority: number,
+		public readonly metadata: PromptMetadata[] = [],
+		public readonly lineBreakBefore: LineBreakBefore,
+		public readonly children: MaterializedNode[],
+		public readonly detail?: 'low' | 'high',
```

Comment:
We should be able to remove the role and children from this element, as they aren't relevant for images.